modules/geochemistry/src/utils/EquilibriumGeochemicalSolver.C

```diff
@@ -0,0 +1,393 @@
+      ss << "Mineral " << _mgd.basis_species_name[swap_out_of_basis]
+         << " consumed.  Swapping it with equilibrium species "
+         << _mgd.eqm_species_name[swap_into_basis] << std::endl;
+      break;
+    }
+  }
+  if (swap_needed)
+    return swap_needed;
+
+  // check maximum saturation index is not positive
+  const std::vector<Real> eqm_SI = _egs.getSaturationIndices();
+  std::vector<unsigned> mineral_SI_order(_num_eqm);
+  ind = 0;
+  std::iota(mineral_SI_order.begin(), mineral_SI_order.end(), ind++);
+  std::sort(mineral_SI_order.begin(), mineral_SI_order.end(), [&](int i, int j) {
+    return eqm_SI[i] > eqm_SI[j];
+  });
+
+  for (const auto & j : mineral_SI_order)
+    if (eqm_SI[j] > 0.0)
+      if (_mgd.eqm_species_mineral[j])
+        if (std::find(_prevent_precipitation.begin(),
+                      _prevent_precipitation.end(),
+                      _mgd.eqm_species_name[j]) == _prevent_precipitation.end())
+        {
+          // mineral has positive saturation index and user is not preventing its precipitation.
+          // determine the basis species to swap out
+          swap_needed = true;
+          swap_into_basis = j;
+          bool legitimate_swap_found = false;
+          swap_out_of_basis = 0;
+          Real best_stoi = 0.0;
+          for (unsigned i = 1; i < _num_basis; ++i) // never swap water (i=0)
+          {
+            if (basis_molality[i] > 0.0 && i != _egs.getChargeBalanceBasisIndex() &&
+                !_mgd.basis_species_gas[i] && !_egs.getBasisActivityKnown()[i])
+            {
+              // don't want to swap out the charge-balance species or any gases of fixed fugacity
+              // or any species with fixed activity
+              const Real stoi = std::abs(_mgd.eqm_stoichiometry(j, i)) / basis_molality[i];
+              if (stoi > best_stoi)
+              {
+                legitimate_swap_found = true;
+                best_stoi = stoi;
+                swap_out_of_basis = i;
+              }
+            }
+          }
+          if (!legitimate_swap_found)
+            mooseException(
+                "Cannot find a legitimate swap for the supersaturated equilibrium species ",
+                _mgd.eqm_species_name[j]);
+          ss << "Mineral " << _mgd.eqm_species_name[j]
+             << " supersaturated.  Swapping it with basis species "
+             << _mgd.basis_species_name[swap_out_of_basis] << std::endl;
+          break;
+        }
+  return swap_needed;
+}
+
+bool
+EquilibriumGeochemicalSolver::reduceInitialResidual()
+{
+  const Real initial_r = _abs_residual;
+  unsigned ind = 0;
+
+  // to get an indication of whether we should increase or decrease molalities in the algorithm
+  // below, find the median of the original molalities
+  const std::vector<Real> original_molality = _egs.getAlgebraicBasisValues();
+  ind = 0;
+  std::vector<unsigned> mol_order(_num_basis_in_algebraic_system);
+  std::iota(mol_order.begin(), mol_order.end(), ind++);
```

Comment:
You do this a ton (7 times) on what seems to be the same type of vectors.

Make a helper for it?

Reply:
Will look at it

Reply:
done